packages/core/src/state-management/state-manager.ts

```diff
@@ -336,6 +336,14 @@ export class StateManager {
     const carFile = await this._buildAnchorRequestCARFile(state$.id, state$.tip)
     const genesisCID = state$.value.log[0].cid
     const genesisCommit = carFile.get(genesisCID)
+
+    // Set the anchor status to PENDING before it is added to the anchor request store.
+    // We do this to prevent the stream from being removed from the anchor request store by previous anchor requests.
```

Comment:
Currently 

1. We request an anchor for CommitA for Stream
2. CommitA is successfully anchored and we receive the anchor cid from pubsub. We have yet to receive the response through the anchor response polling 
3. We create CommitB (anchor status NOT_REQUESTED) and anchor it. When we anchor a stream we add it to the request store before `_processAnchorResponse` is called. `_processAnchorResponse` is what updates the Stream's anchor status to PENDING 
4. We receive the anchor response for CommitA. We remove Stream from the anchor request store. There is nothing in the anchor request store for CommitB

We should have something in the anchor request store for CommitB so
4. We receive the anchor response for CommitA. Since CommitA is not the tip we do not remove Stream from the anchor request store.

Unfortunately we cannot rely on only the tip, because there could be a situation where the tip does not have an outstanding anchor request

4. We receive the anchor response for CommitA. Since its not the tip, and the tip has a requested anchor (the anchor status is PENDING) we do not remove the stream from the store

BUT because we could be receiving the anchor response after we have added the stream to the anchor request store but before the anchor status of the Stream has been updated to PENDING via `_processAnchorResponse` we could still erroneously remove the Stream from the anchor request store. 

We could only remove things from the store if the stream and cid matches, but our current stores do not support atomic transactions (a read + write in one go). 

So the solution I have ended with is updating the anchor status to PENDING before we add it to the anchor request store. We could get into a situation where the anchor status has been updated to PENDING but things crash before we add the stream to the anchor request store but i'm not sure what to do there.

Reply:
okay, I *think* I get it.

If we swap the order of writing to the persistent anchor request store and updating the state to PENDING, does that address the race condition you mention at the end where the node crashes before the stream is added to the anchor request store?

Reply:
No it does not. 

We really need some sort of db transaction here.